### PR TITLE
Add a constructor for TouchEvent and Touch

### PR DESCRIPTION
**Pull request by RByers**
As previously resolved by the web events working group.  If we're going to have a Level 2 spec, then we should do this.

### CODE REVIEW

**Comment by RByers** [comment]
Note that you can preview the result of this [here](https://rawgit.com/RByers/touch-events/constructor/touchevents.html#touch-interface).

**Comment by RByers** [comment]
A couple areas to pay attention to: 

The EventTarget isn't supposed to be nullable, so we can't have a default value - so I made it required (along with the `identifier` for good measure).  But note that in blink at least, it is actually nullable - `document.createTouch()` will create a `Touch` object with a `null` target).

Like for `createTouch` I didn't allow `clientX` and `clientY` to be explicitly initialized.  Instead they're initialized implicitly by `pageX` and `pageY`.  This is similar to the `MouseEvent` constructor, except there it's the client co-ordinates that are explicit with the page ones computed from them.  Perhaps it's better to follow the `MouseEvent` pattern?

I initialize the `TouchList` objects with a `sequence<Touch>` (not unlike [the FontFaceSetLoadEvent constructor](http://dev.w3.org/csswg/css-font-loading/#dictdef-fontfacesetloadeventinit)).

File: touchevents.html [hunk 1]
```diff
@@ -440,6 +463,23 @@ <h3 id="webidl-conform">WebIDL Conformance</h3>
         </dd>
       </dl>
 
+      <dl title='dictionary TouchEventInit : UIEventInit' class='idl'>
+        <dt>sequence&lt;Touch&gt; touches = []</dt>
+        <dd>Initializes the <code>touches</code> property of the <code>TouchEvent</code> object with each element from the sequence.</dd>
+        <dt>sequence&lt;Touch&gt; targetTouches = []</dt>
+        <dd>Initializes the <code>touches</code> property of the <code>TouchEvent</code> object with each element from the sequence.</dd>
```
Comment (mbrubeck) [inline touchevents.html]: typo: `touches` should be `targetTouches`

File: touchevents.html [hunk 2]
```diff
@@ -440,6 +463,23 @@ <h3 id="webidl-conform">WebIDL Conformance</h3>
...
+        <dt>sequence&lt;Touch&gt; changedTouches = []</dt>
+        <dd>Initializes the <code>touches</code> property of the <code>TouchEvent</code> object with each element from the sequence.</dd>
```
Comment (mbrubeck) [inline touchevents.html]: Typo: `touches` should be `changedTouches`

**Comment by mbrubeck** [comment]
> Like for createTouch I didn't allow clientX and clientY to be explicitly initialized. Instead they're initialized implicitly by pageX and pageY. This is similar to the MouseEvent constructor, except there it's the client co-ordinates that are explicit with the page ones computed from them. Perhaps it's better to follow the MouseEvent pattern?

Following the MouseEvent pattern would have the advantage of making it easier to port code from MouseEvent to TouchEvent (or vice-versa).  Other than that, I have no strong feelings about this.

**Comment by RByers** [comment]
> Following the MouseEvent pattern would have the advantage of making it easier to port code from MouseEvent to TouchEvent (or vice-versa). Other than that, I have no strong feelings about this.

Right, and also between PointerEvent and TouchEvent too.

Personally I also think of the "client" co-ordinates as being the primary ones and the "page" ones being synthesized from that (at least that's how our implementation works).

I don't see much value in matching exactly the form of document.createTouch - [our data](https://www.chromestatus.com/metrics/feature/popularity#V8Document_CreateTouch_Method) suggests that that has extremely low usage in the wild.

That seems like enough reason to change to me.  I'll update the PR with that change soon.

File: touchevents.html (hunk 1 above)
Comment (RByers) [inline touchevents.html]: Done, thanks.

File: touchevents.html (hunk 2 above)
Comment (RByers) [inline touchevents.html]: Fixed, thanks.

**Comment by RByers** [comment]
Ok, how does this look now?

**Comment by mbrubeck** [comment]
Looks good to me.  Free free to merge this, unless you want to wait for any other feedback.

**Comment by RByers** [comment]
Thanks.  I'll go ahead and merge now, but happy to take additional feedback post-merge.

**Comment by teddink** [comment]
Jacob Rossi and I chatted about this - we agree with the approach.

**Comment by RByers** [comment]
@teddink Excellent, thank you!

**Pull request merged**